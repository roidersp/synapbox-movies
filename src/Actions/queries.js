import gql from "graphql-tag";

export const ADD_TO_CART = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) @client
  }
`;

export const DELETE_FROM_CART = gql`
  mutation deleteFromCart($id: ID!) {
    deleteFromCart(id: $id) @client
  }
`;

export const GET_CART_ITEMS = gql`
  query GetCartItems {
    cartItems @client
  }
`;

export const GET_MOVIE_ID = gql`
  {
    movieId @client
  }
`;

export const GET_MOVIE = gql`
  query($id: ID!) {
    movie: item(where: { id: $id }) {
      id
      title
      description
      largeImage
      price
    }
  }
`;

export const IS_IN_CART = gql`
  query isInCart($id: ID!) {
    isInCart(id: $id) @client
  }
`;

export const CART_IS_OPEN = gql`
  {
    cartIsOpen @client
    cartItems @client
  }
`;

export const CONFIRM_DELETE = gql`
  {
    dialogDelete @client {
      open
      movie {
        id
        title
        image
      }
    }
  }
`;

export const HANDLE_CONFIRM_DELETE = gql`
  {
    dialogDelete @client {
      open
      Movie
    }
  }
`;

export const GET_MOVIES = gql`
  {
    movies: itemsConnection(where: null, first: 16, after: null) {
      aggregate {
        count
      }
      edges {
        node {
          id
          title
          description
          image
          price
        }
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
  }
`;

export const GET_MOVIES_BY_ID = gql`
  query search($ids: [ID!]) {
    movies: items(where: { id_in: $ids }) {
      id
      title
      image
      price
    }
  }
`;

export const GET_SEARCH_MOVIES = gql`
  query search($searchValue: String!) {
    movies: itemsConnection(
      where: { title_contains: $searchValue }
      first: 5
      after: null
    ) {
      edges {
        node {
          id
          title
          description
          image
          price
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
`;
