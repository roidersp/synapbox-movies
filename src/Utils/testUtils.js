import wait from "waait";
import { act } from "@testing-library/react";

export const actWait = async (amount = 0) => {
  await act(async () => {
    await wait(amount);
  });
};
