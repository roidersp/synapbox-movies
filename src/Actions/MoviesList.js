import React from "react";

import gql from "graphql-tag";
import { Query } from "react-apollo";

import Grid from "@material-ui/core/Grid";

import MovieCard from "../Components/MovieCard";
import SearchBox from "../Components/SearchBox";

const GET_MOVIES = gql`
  {
    movies: itemsConnection(where: null, first: 24, after: null) {
      aggregate {
        count
      }
      edges {
        node {
          id
          title
          description
          image
          price
        }
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
  }
`;

const MoviesList = ({ action }) => (
  <Query query={GET_MOVIES}>
    {({ loading, error, data }) => {
      if (loading) return "Loading...";
      if (error) return `Error! ${error.message}`;

      const { edges: movies } = data.movies;

      return (
        <Grid container spacing={2}>
          <Grid item xs={12}>
            <SearchBox action={action} />
          </Grid>
          {movies.map(({ node: movie }, index) => (
            <MovieCard {...movie} key={index} action={action} />
          ))}
        </Grid>
      );
    }}
  </Query>
);

export default MoviesList;
