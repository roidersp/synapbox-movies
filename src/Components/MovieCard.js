import React from "react";
import { useMutation, useQuery } from "@apollo/react-hooks";

import { makeStyles, withStyles } from "@material-ui/core/styles";
import {
  Card,
  CardActions,
  CardContent,
  CardMedia,
  Typography,
  Grid,
  Fab
} from "@material-ui/core";

import { ADD_TO_CART, IS_IN_CART } from "../Actions/queries";
import { AddShoppingCart, Visibility } from "@material-ui/icons";

const useStyles = makeStyles({
  card: {
    width: "100%",
    height: "100%",
    display: "flex",
    flexDirection: "column",
    alignItems: "flex-end",
    justifyContent: "space-between"
  },
  media: {
    height: 160,
    width: "100%",
    flex: "0"
  },
  addButton: {
    fontSize: 18
  },
  seeButton: {
    marginRight: 10
  },
  title: {
    marginTop: 10
  },
  cardActions: {
    display: "flex",
    justifyContent: "space-between",
    marginBottom: 15
  }
});

const CustomIcon = withStyles(theme => ({
  root: {
    color: "#fff",
    backgroundColor: theme.palette.secondary.dark,
    padding: 8,
    "&:hover": {
      backgroundColor: theme.palette.secondary.light,
      color: "#c3c3c3"
    }
  }
}))(Fab);

const MovieCard = ({ id, title, image, price }) => {
  const classes = useStyles();

  // const {data, loading: queryLoading , error: queryError, ...lets} = useQuery(
  //   IS_IN_CART,
  //   { variables: { id: id } }
  // )

  const [mutate, { loading, error, client, called, ...rest }] = useMutation(
    ADD_TO_CART,
    {
      variables: { id: id }
      // refetchQueries: [
      //   {
      //     query: IS_IN_CART,
      //     variables: { id: id },
      //   }
      // ]
    }
  );

  const setMovie = () => {
    client.writeData({ data: { movieId: id } });
  };

  if (loading) console.log("loading mutation");

  if (error) console.log("error", error);

  if (called) {
    console.log("called");
  }

  return (
    <Grid item lg={3} md={4} sm={6} xs={12}>
      <Card className={classes.card}>
        <div style={{ width: "100%" }}>
          <CardMedia className={classes.media} image={image} title={title} />
          <CardContent>
            <Typography className={classes.title} variant="h5" component="h2">
              {title}
            </Typography>
            <Typography variant="body2" color="textSecondary" component="p">
              ${price}
            </Typography>
          </CardContent>
        </div>
        <CardActions className={classes.cardActions}>
          <Fab variant="extended" size="medium" onClick={setMovie}>
            <Visibility className={classes.seeButton} />
            Ver detalle
          </Fab>
          <CustomIcon variant="extended" size="medium" onClick={mutate}>
            {loading ? (
              <p>Loading...</p>
            ) : error ? (
              <p>Ocurrio un error</p>
            ) : (
              <AddShoppingCart className={classes.addButton} />
            )}
          </CustomIcon>
        </CardActions>
      </Card>
    </Grid>
  );
};

export default MovieCard;
