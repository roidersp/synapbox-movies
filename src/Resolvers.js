import gql from "graphql-tag";
import { GET_CART_ITEMS } from "./Actions/queries";

export const resolvers = {
  Mutation: {
    addToCart: (_, { id }, { cache }) => {
      const { cartItems } = cache.readQuery({ query: GET_CART_ITEMS });

      const data = {
        cartItems: !cartItems.includes(id) ? [...cartItems, id] : cartItems
      };

      cache.writeQuery({ query: GET_CART_ITEMS, data });
      return data.cartItems;
    },
    deleteFromCart: (_, { id }, { cache }) => {
      const { cartItems } = cache.readQuery({ query: GET_CART_ITEMS });

      const data = {
        cartItems: cartItems.includes(id)
          ? cartItems.filter(i => i !== id)
          : cartItems
      };

      cache.writeQuery({ query: GET_CART_ITEMS, data });
      return cartItems.includes(id);
    }
  }
};

export const typeDefs = `
  extend type Mutation {
    addToCart(id: ID!)
    deleteFromCart(id: ID!)
  }
`;
