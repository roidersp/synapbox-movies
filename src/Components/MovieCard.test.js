import React from "react";
// import {shallow,mount,render} from 'enzyme';
import { MockedProvider } from "@apollo/react-testing";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { actWait } from "../Utils/testUtils";
import { ADD_TO_CART, GET_CART_ITEMS, GET_MOVIE_ID } from "../Actions/queries";
import MovieCard from "./MovieCard";
import { InMemoryCache } from "apollo-cache-inmemory";
import { resolvers } from "../Resolvers";

const mocks = [
  {
    request: {
      query: ADD_TO_CART,
      variables: { id: "ck49086bqsnup0b09d14atpb2" }
    },
    newData: jest.fn(() => ({
      data: {
        cartItems: ["ck49086bqsnup0b09d14atpb2"]
      }
    }))
  }
];

const mockError = [
  {
    request: {
      query: ADD_TO_CART,
      variables: { id: "ck49086bqsnup0b09d14atpb2" }
    },
    result: {
      errors: [new Error("error occurred")]
    }
  }
];

const data = {
  image: "http://static.tvmaze.com/uploads/images/medium_landscape/1/4388.jpg",
  description:
    "When the residents of Chester's Mill find themselves trapped under a massive transparent dome with no way out, they struggle to survive as resources rapidly dwindle and panic quickly escalates.",
  price: 500,
  id: "ck49086bqsnup0b09d14atpb2",
  title: "Pilot"
};

afterEach(cleanup);

describe("<MovieCard>", () => {
  it("snapshot MovieCard", async () => {
    const Wrapper = render(
      <MockedProvider mocks={mocks} addTypename={false} resolvers={{}}>
        <MovieCard {...data} />
      </MockedProvider>
    );
    await actWait();
    expect(Wrapper).toMatchSnapshot();
  });

  it("Expect GridContainer and Card", async () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypename={false} resolvers={{}}>
        <MovieCard {...data} />
      </MockedProvider>
    );

    await actWait();

    const Grid = getByTestId("GridCard");
    expect(Grid).toBeInTheDocument();

    const Card = getByTestId("Card");
    expect(Card).toBeInTheDocument();
  });

  it("Add Movie to cart success", async () => {
    const cache = new InMemoryCache();

    cache.writeData({
      data: {
        cartItems: []
      }
    });

    const { getByTestId } = render(
      <MockedProvider
        mocks={mocks}
        addTypename={false}
        resolvers={resolvers}
        cache={cache}
      >
        <MovieCard {...data} />
      </MockedProvider>
    );

    fireEvent.click(getByTestId("addCartButton"));

    await actWait();

    const { cartItems } = cache.readQuery({ query: GET_CART_ITEMS });
    expect(cartItems).toContain("ck49086bqsnup0b09d14atpb2");
    expect(cartItems).toHaveLength(1);
  });

  it("Click View movie deatils", async () => {
    const cache = new InMemoryCache();

    const { getByTestId } = render(
      <MockedProvider
        mocks={mocks}
        addTypename={false}
        resolvers={{}}
        cache={cache}
      >
        <MovieCard {...data} />
      </MockedProvider>
    );

    fireEvent.click(getByTestId("ViewDetails"));

    await actWait();

    const { movieId } = cache.readQuery({ query: GET_MOVIE_ID });

    expect(movieId).toBe("ck49086bqsnup0b09d14atpb2");
  });

  it("Get Loading text", async () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypename={false} resolvers={{}}>
        <MovieCard {...data} />
      </MockedProvider>
    );

    fireEvent.click(getByTestId("addCartButton"));

    const loading = getByTestId("loading");
    expect(loading).toBeInTheDocument();
  });
});
