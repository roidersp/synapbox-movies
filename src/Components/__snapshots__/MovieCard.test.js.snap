// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard> snapshot MovieCard 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
        data-testid="GridCard"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="Card"
        >
          <div
            style="width: 100%;"
          >
            <div
              class="MuiCardMedia-root makeStyles-media-2"
              style="background-image: url(http://static.tvmaze.com/uploads/images/medium_landscape/1/4388.jpg);"
              title="Pilot"
            />
            <div
              class="MuiCardContent-root"
            >
              <h2
                class="MuiTypography-root makeStyles-title-5 MuiTypography-h5"
              >
                Pilot
              </h2>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
              >
                $
                500
              </p>
            </div>
          </div>
          <div
            class="MuiCardActions-root makeStyles-cardActions-6 MuiCardActions-spacing"
          >
            <button
              class="MuiButtonBase-root MuiFab-root MuiFab-extended MuiFab-sizeMedium"
              data-testid="ViewDetails"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiFab-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-seeButton-4"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                  />
                </svg>
                Ver detalle
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiFab-root WithStyles(ForwardRef(Fab))-root-197 MuiFab-extended MuiFab-sizeMedium"
              data-testid="addCartButton"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiFab-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-addButton-3"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M0 0h24v24H0zm18.31 6l-2.76 5z"
                    fill="none"
                  />
                  <path
                    d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
      data-testid="GridCard"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
        data-testid="Card"
      >
        <div
          style="width: 100%;"
        >
          <div
            class="MuiCardMedia-root makeStyles-media-2"
            style="background-image: url(http://static.tvmaze.com/uploads/images/medium_landscape/1/4388.jpg);"
            title="Pilot"
          />
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root makeStyles-title-5 MuiTypography-h5"
            >
              Pilot
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              $
              500
            </p>
          </div>
        </div>
        <div
          class="MuiCardActions-root makeStyles-cardActions-6 MuiCardActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiFab-root MuiFab-extended MuiFab-sizeMedium"
            data-testid="ViewDetails"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiFab-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-seeButton-4"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                />
              </svg>
              Ver detalle
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiFab-root WithStyles(ForwardRef(Fab))-root-197 MuiFab-extended MuiFab-sizeMedium"
            data-testid="addCartButton"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiFab-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-addButton-3"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0 0h24v24H0zm18.31 6l-2.76 5z"
                  fill="none"
                />
                <path
                  d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
